main.cc

```diff
@@ -1724,11 +1724,6 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             auto stop_repair_service = defer_verbose_shutdown("repair service", [&repair] {
```

Comment:
Up above there's sharded<repair_service>::start() call, so the start sequence is broken into parts. If moving repair_service::start() calls lower, the sharded<repair_service>::start() should be moved lower too

Reply:
But it can't. Between `repair_service::start` and `batchlog_manager::start` we have `storage_service::join_cluster`. It can't be called before `repair_service::start`, since it uses `repair_service`. It can't be used after `batchlog_manager::start` since `token_metadata_impl::_normal_token_owners` won't be updated (which is normally done by `update_topology_change_info` in `storage_service::bootstrap`) and so we will have division by 0 in `batchlog_manager::replay_all_failed_batches()`

So we have:
repair_service::start -> join_cluster -> batchlog_manager::start -> repair_service::start
But repair_service uses bm only in message handler

Reply:
> But it can't. Between repair_service::start and batchlog_manager::start we have storage_service::join_cluster. It can't be called before repair_service::start, since it uses repair_service. It can't be used after batchlog_manager::start because batchlog_manager::replay_all_failed_batches() needs to have token_metadata_impl::_normal_token_owners updated, which is normally done by update_topology_change_info in storage_service::bootstrap.

Nice circular dependency, but still it shouldn't be cut by splitting repair service initialization split. If storage service needs repair_service to be usable by the time join_cluster() is called, then repair_service::start() must have been called by that time, it's not correct to delay it. You say that join_cluster() needs repair. After this patch repair_service is in a state _without_ handlers at all. Why is that OK that storage service uses repair_service without handlers?

Reply:
Right. So I think the best solution will be to throw in `repair_service::repair_flush_hints_batchlog_handler` if batchlog manager isn't initialized. That won't cause repair to fail